### PR TITLE
Issue#1380 create notifications

### PR DESCRIPTION
**Pull request by Hugo-Hontans**
- [x] Create notifications for:
    - [x] orgName: notif related to org (new members & request)
        - [x] Org refuse user request to join
        - [x] User refuse org invitation
        - [x] Remove a member from org
        - [x] Add a new member to org
    - [x] Archipel: org acceptation by Archipel Content
    - [x] Titles: CRUD on movies

Contributes to #1380

### CODE REVIEW

File: apps/backend-functions/src/movie.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import { functions, db } from './internals/firebase';
+import { MovieDocument, OrganizationDocument, PublicUser } from './data/types';
+import { createNotification, NotificationType } from '@blockframes/notification/types';
+import { App } from '@blockframes/utils/apps';
+import { triggerNotifications } from './notification';
+import { flatten, isEqual } from 'lodash';
+
+/** Create a notification with user and movie. */
+function notifUser(userId: string, notificationType :NotificationType, movie: MovieDocument, user: PublicUser) {
```
Comment (lsenta) [inline apps/backend-functions/src/movie.ts]: autoformat

File: apps/backend-functions/src/orgs.ts [hunk 2]
```diff
@@ -12,6 +12,48 @@ import { RelayerConfig, relayerDeployOrganizationLogic, relayerRegisterENSLogic,
...
+    },
+    app: App.blockframes
+  });
+}
+
+/** Send notifications to all org's members when a member is added or removed. */
+async function sendNotificationToUsers(before: OrganizationDocument, after: OrganizationDocument) {
+  // Member added
+  if (before.userIds.length < after.userIds.length) {
+    const userAddedId = after.userIds.filter(userId => !before.userIds.includes(userId))[0];
```
Comment (lsenta) [inline apps/backend-functions/src/orgs.ts]: hard to understand, and prone to error (mutliple add for example).
Use an explicit difference https://lodash.com/docs/#difference

File: apps/backend-functions/src/orgs.ts [hunk 3]
```diff
@@ -12,6 +12,48 @@ import { RelayerConfig, relayerDeployOrganizationLogic, relayerRegisterENSLogic,
...
+    const userSnapshot = await db.doc(`users/${userAddedId}`).get();
+    const userAdded = userSnapshot.data() as PublicUser;
+
+    const notifications = after.userIds.map(userId => notifUser(userId, NotificationType.memberAddedToOrg, after, userAdded));
+    return triggerNotifications(notifications);
+
+  // Member removed
+  } else if (before.userIds.length > after.userIds.length) {
+    const userAddedId = before.userIds.filter(userId => !after.userIds.includes(userId))[0];
```
Comment (lsenta) [inline apps/backend-functions/src/orgs.ts]: Same https://lodash.com/docs/#difference

File: apps/backend-functions/src/orgs.ts [hunk 4]
```diff
@@ -12,6 +12,48 @@ import { RelayerConfig, relayerDeployOrganizationLogic, relayerRegisterENSLogic,
...
+    organization: {
+      id: org.id,
+      name: org.name
+    },
+    app: App.blockframes
+  });
+}
+
+/** Send notifications to all org's members when a member is added or removed. */
+async function sendNotificationToUsers(before: OrganizationDocument, after: OrganizationDocument) {
```
Comment (lsenta) [inline apps/backend-functions/src/orgs.ts]: Rename, this does not "send notification to users", it "notify org member of users added and removed".

File: apps/backend-functions/src/orgs.ts [hunk 5]
```diff
@@ -54,6 +96,9 @@ export async function onOrganizationUpdate(
     throw new Error('Organization name cannot be changed !'); // this will require to change the org ENS name, for now we throw to prevent silent bug
   }
 
+  // Send notifications when a member is added or removed
+  await sendNotificationToUsers(before, after);
```
Comment (lsenta) [inline apps/backend-functions/src/orgs.ts]: No idea what this code does, rename the function, the comment should become useless.

File: apps/backend-functions/src/orgs.ts [hunk 6]
```diff
@@ -63,6 +108,12 @@ export async function onOrganizationUpdate(
   if (becomeAccepted) {
     // send email to let the org admin know that the org has been accepted
     await sendMailFromTemplate(organizationWasAccepted(admin.email, id, admin.name));
+    const notification = createNotification({
+      userId: after.userIds[0],
```
Comment (lsenta) [inline apps/backend-functions/src/orgs.ts]: `[0]` seems arbitrary, what are you trying to do?

File: libs/organization/src/lib/+state/organization.firestore.ts [hunk 7]
```diff
@@ -32,7 +32,8 @@ export interface OrganizationDocumentWithDates extends OrganizationRaw<Date> {}
 /** Status of an Organization, set to pending by default when an Organization is created. */
 export const enum OrganizationStatus {
   pending = 'pending',
-  accepted = 'accepted'
+  accepted = 'accepted',
+  refused = 'refused'
```
Comment (lsenta) [inline libs/organization/src/lib/+state/organization.firestore.ts]: be consistent, use either declined or refused but not both.

File: libs/notification/src/lib/notification/+state/notification.firestore.ts [hunk 8]
```diff
@@ -11,22 +12,32 @@ export const enum NotificationType {
   deleteDocument = 'deleteDocument',
   inviteOrganization = 'inviteOrganization',
   removeOrganization = 'removeOrganization',
-  pathToDocument = 'pathToDocument'
+  pathToDocument = 'pathToDocument',
+  // Organization acceptation by Archipel Content
+  organizationAccepted = 'organizationAccepted',
```
Comment (lsenta) [inline libs/notification/src/lib/notification/+state/notification.firestore.ts]: If you need to document the name, maybe the name is not good.

File: libs/movie/src/lib/movie/+state/movie.service.ts [hunk 9]
```diff
@@ -42,16 +46,22 @@ export class MovieService extends CollectionService<MovieState> {
     )
   }
 
+  public async removeMovie(movieId: string) {
```
Comment (lsenta) [inline libs/movie/src/lib/movie/+state/movie.service.ts]: Looks like a trap for other dev and futur you.
They won't remember nor understand which function to call. Don't declare multiple methods that do pretty much the same thing with pretty much the same name.

Why do you need this?

File: apps/delivery/delivery/src/app/home/home.component.ts [hunk 10]
```diff
@@ -35,7 +35,7 @@ export class DeliveryHomeComponent implements OnInit {
 
   public remove(movie: Movie) {
     try {
-      this.movieService.remove(movie.id);
+      this.movieService.removeMovie(movie.id);
```
Comment (lsenta) [inline apps/delivery/delivery/src/app/home/home.component.ts]: what's wrong with remove?

File: libs/movie/src/lib/movie/+state/movie.service.ts (hunk 9 above)
Comment (lsenta) [reply libs/movie/src/lib/movie/+state/movie.service.ts]: Involve @GrandSchtroumpf for the frontend part

File: apps/delivery/delivery/src/app/home/home.component.ts (hunk 10 above)
Comment (Hugo-Hontans) [reply apps/delivery/delivery/src/app/home/home.component.ts]: Need this to update the movie with the userId before delete it

File: apps/backend-functions/src/orgs.ts (hunk 6 above)
Comment (Hugo-Hontans) [reply apps/backend-functions/src/orgs.ts]: The organization has just been created, so we are sure that there is only one userId in the array

Comment (lsenta) [reply apps/backend-functions/src/orgs.ts]: Add a comment

File: apps/backend-functions/src/orgs.ts [hunk 11]
```diff
@@ -12,6 +14,48 @@ import { RelayerConfig, relayerDeployOrganizationLogic, relayerRegisterENSLogic,
...
+    const userAddedId = difference(after.userIds, before.userIds)[0];
+    const userSnapshot = await db.doc(`users/${userAddedId}`).get();
+    const userAdded = userSnapshot.data() as PublicUser;
+
+    const notifications = after.userIds.map(userId => notifUser(userId, NotificationType.memberAddedToOrg, after, userAdded));
+    return triggerNotifications(notifications);
+
+  // Member removed
+  } else if (before.userIds.length > after.userIds.length) {
+    const userAddedId = difference(before.userIds, after.userIds)[0];
```
Comment (lsenta) [inline apps/backend-functions/src/orgs.ts]: userRemovedID

File: apps/backend-functions/src/orgs.ts [hunk 12]
```diff
@@ -12,6 +14,48 @@ import { RelayerConfig, relayerDeployOrganizationLogic, relayerRegisterENSLogic,
...
+    organization: {
+      id: org.id,
+      name: org.name
+    },
+    app: App.blockframes
+  });
+}
+
+/** Send notifications to all org's members when a member is added or removed. */
+async function sendNotificationToMembers(before: OrganizationDocument, after: OrganizationDocument) {
```
Comment (lsenta) [inline apps/backend-functions/src/orgs.ts]: Rename `notifOnOrgMemberChanges`

File: libs/movie/src/lib/movie/+state/movie.firestore.ts [hunk 13]
```diff
@@ -196,10 +196,17 @@ interface MovieSalesInfoRaw<D> {
...
+  createdBy: string;
+  updatedBy?: string,
+}
+
 /** Generic interface of a Movie */
 interface MovieRaw<D> {
   // @todo #643 add new fields to Draw.io
   _type: 'movies';
+  _meta?: DocumentMeta;
+  processedId?: string;
```
Comment (lsenta) [inline libs/movie/src/lib/movie/+state/movie.firestore.ts]: Do you use it in the end?

File: libs/notification/src/lib/notification/+state/notification.firestore.ts [hunk 14]
```diff
@@ -11,22 +12,31 @@ export const enum NotificationType {
   deleteDocument = 'deleteDocument',
   inviteOrganization = 'inviteOrganization',
   removeOrganization = 'removeOrganization',
-  pathToDocument = 'pathToDocument'
+  pathToDocument = 'pathToDocument',
+  organizationAcceptedByArchipelContent = 'organizationAcceptedByArchipelContent',
```
Comment (lsenta) [inline libs/notification/src/lib/notification/+state/notification.firestore.ts]: 👍

File: libs/movie/src/lib/movie/+state/movie.firestore.ts (hunk 13 above)
Comment (Hugo-Hontans) [reply libs/movie/src/lib/movie/+state/movie.firestore.ts]: Yes sure, when a user create a movie and when a user update a movie, we update the field createdBy or updatedBy de `_meta`

File: apps/backend-functions/src/orgs.ts (hunk 6 above)
Comment (Hugo-Hontans) [reply apps/backend-functions/src/orgs.ts]: sure

File: apps/backend-functions/src/main.ts [hunk 15]
```diff
@@ -117,6 +118,26 @@ export const onInvitationUpdateEvent = onDocumentWrite(
...
+/**
+ * Trigger: when a movie is updated
+ */
+export const onMovieUpdateEvent = onDocumentUpdate(
+  'movies/{movieId}',
+  onMovieUpdate
+)
+
+/** Trigger: REST call to remove a movie and send notifications for all users of concerned organizations. */
+export const removeMovie = functions.https
```
Comment (lsenta) [inline apps/backend-functions/src/main.ts]: Don't do that, you lost auth & security, use a regular trigger

File: libs/movie/src/lib/movie/+state/movie.service.ts [hunk 16]
```diff
@@ -42,16 +46,26 @@ export class MovieService extends CollectionService<MovieState> {
     )
   }
 
+  /** Update deletedBy (_meta field of movie) with the current user and remove the movie. */
+  public async remove(movieId: string) {
+    const userId = this.authQuery.userId;
+    const meta = this.query.getEntity(movieId)._meta;
+    // We need to update the _meta field before remove to get the userId in the backend function: onMovieDeleteEvent
+    await this.db.doc(`movies/${movieId}`).update({ _meta: { ...meta, deletedBy: userId } });
```
Comment (lsenta) [inline libs/movie/src/lib/movie/+state/movie.service.ts]: fwiw I think you could do:
```
this.db....update({"_meta.deleteBy": userId});
```

https://firebase.google.com/docs/firestore/manage-data/add-data#update_fields_in_nested_objects

That'll spare you the `...meta` thingy

File: libs/movie/src/lib/movie/+state/movie.service.ts [hunk 17]
```diff
@@ -60,6 +74,12 @@ export class MovieService extends CollectionService<MovieState> {
     return movie;
   }
 
+  onUpdate(form: Movie, { write }: WriteOptions) {
+    const movie = this.query.getActive();
+    const movieRef = this.db.doc(`movies/${movie.id}`).ref;
+    write.update(movieRef, { _meta: { ...movie._meta, updatedBy: this.authQuery.userId } });
```
Comment (lsenta) [inline libs/movie/src/lib/movie/+state/movie.service.ts]: same https://firebase.google.com/docs/firestore/manage-data/add-data#update_fields_in_nested_objects

File: apps/backend-functions/src/invitation.ts [hunk 18]
```diff
@@ -107,6 +114,33 @@ async function onInvitationToOrgAccept({ user, organization }: InvitationFromOrg
...
+  const org = orgSnapshot.data() as OrganizationDocument;
+
+  const userSnapshot = await db.doc(`users/${invitation.user.uid}`).get();
+  const user = userSnapshot.data() as PublicUser;
+
+  const permissionSnapshot = await db.doc(`permissions/${invitation.organization.id}`).get();
+  const permission = permissionSnapshot.data()!;
+
+  const isAdmin = (userId: string) => (permission.roles[userId] === UserRole.admin) || (permission.roles[userId] === UserRole.superAdmin);
+  const adminIds = org.userIds.filter(isAdmin);
```
Comment (lsenta) [inline apps/backend-functions/src/invitation.ts]: Could you use:
`const adminIds = await getAdminIds(organizationId);`

and remove lines 119 -> 128?

(it's used in a few places, for example: `mailOrganizationAdminOnAccept`)

File: apps/backend-functions/src/invitation.ts [hunk 19]
```diff
@@ -277,6 +314,30 @@ async function onInvitationFromUserToJoinOrgAccept({
...
+/** Send a notification to admins of organization to notify them that the request is declined. */
+async function onInvitationFromUserToJoinOrgDecline(invitation: InvitationFromUserToOrganization) {
+  const orgSnapshot = await db.doc(`orgs/${invitation.organization.id}`).get();
+  const org = orgSnapshot.data() as OrganizationDocument;
+
+  const permissionSnapshot = await db.doc(`permissions/${invitation.organization.id}`).get();
+  const permission = permissionSnapshot.data()!;
+
+  const isAdmin = (userId: string) => (permission.roles[userId] === UserRole.admin) || (permission.roles[userId] === UserRole.superAdmin);
+  const adminIds = org.userIds.filter(isAdmin);
```
Comment (lsenta) [inline apps/backend-functions/src/invitation.ts]: `const admins = await getAdminIds(organizationId);`

File: libs/movie/src/lib/movie/+state/movie.firestore.ts (hunk 13 above)
Comment (lsenta) [reply libs/movie/src/lib/movie/+state/movie.firestore.ts]: That comment is on `processedId` :0

File: libs/movie/src/lib/movie/+state/movie.service.ts (hunk 16 above)
Comment (Hugo-Hontans) [reply libs/movie/src/lib/movie/+state/movie.service.ts]: Wtf, I did't know that

File: libs/movie/src/lib/movie/+state/movie.firestore.ts (hunk 13 above)
Comment (Hugo-Hontans) [reply libs/movie/src/lib/movie/+state/movie.firestore.ts]: ....

**Pull request merged**